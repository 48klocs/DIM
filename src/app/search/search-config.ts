import { DestinyVersion } from '@destinyitemmanager/dim-api-types';
import { destinyVersionSelector } from 'app/accounts/selectors';
import { createSelector } from 'reselect';
import { FilterDefinition } from './filter-types';
import type { QueryAST } from './query-parser';
import advancedFilters from './search-filters/advanced';
import d1Filters from './search-filters/d1-filters';
import dupeFilters from './search-filters/dupes';
import freeformFilters from './search-filters/freeform';
import itemInfosFilters from './search-filters/item-infos';
import knownValuesFilters from './search-filters/known-values';
import loadoutFilters from './search-filters/loadouts';
import simpleRangeFilters from './search-filters/range-numeric';
import overloadedRangeFilters from './search-filters/range-overload';
import simpleFilters from './search-filters/simple';
import socketFilters from './search-filters/sockets';
import statFilters from './search-filters/stats';
import locationFilters from './search-filters/stores';
import wishlistFilters from './search-filters/wishlist';

const allFilters = [
  ...dupeFilters,
  ...($featureFlags.wishLists ? wishlistFilters : []),
  ...freeformFilters,
  ...itemInfosFilters,
  ...knownValuesFilters,
  ...d1Filters,
  ...loadoutFilters,
  ...simpleRangeFilters,
  ...overloadedRangeFilters,
  ...simpleFilters,
  ...socketFilters,
  ...statFilters,
  ...locationFilters,
  ...advancedFilters,
];

export const searchConfigSelector = createSelector(destinyVersionSelector, buildSearchConfig);

//
// SearchConfig
//

export interface SearchConfig {
  allFilters: FilterDefinition[];
  filters: Record<string, FilterDefinition>;
  keywords: string[];
}

/** Builds an object that describes the available search keywords and filter definitions. */
export function buildSearchConfig(destinyVersion: DestinyVersion): SearchConfig {
  const keywords = new Set<string>();
  const allFiltersByKeyword: Record<string, FilterDefinition> = {};
  const allApplicableFilters: FilterDefinition[] = [];
  for (const filter of allFilters) {
    if (!filter.destinyVersion || filter.destinyVersion === destinyVersion) {
      for (const keyword of generateSuggestionsForFilter(filter)) {
        keywords.add(keyword);
      }
      allApplicableFilters.push(filter);
      const filterKeywords = Array.isArray(filter.keywords) ? filter.keywords : [filter.keywords];
      for (const keyword of filterKeywords) {
        allFiltersByKeyword[keyword] = filter;
      }
    }
  }

  return {
    allFilters: allApplicableFilters,
    keywords: Array.from(keywords),
    filters: allFiltersByKeyword,
  };
}

/**
 * loops through collections of strings (filter segments), generating combinations
 *
 * for example, with
 * `[ [a], [b,c], [d,e] ]`
 * as an input, this generates
 *
 * `[ a:, a:b:, a:c:, a:b:d, a:b:e, a:c:d, a:c:e ]`
 */
function expandStringCombinations(stringGroups: string[][], minDepth = 0) {
  const results: string[][] = [];
  for (let i = 0; i < stringGroups.length; i++) {
    const stringGroup = stringGroups[i];
    const stems = results.length ? results[results.length - 1] : undefined;
    const newResults = stringGroup.flatMap((suffix) =>
      stems
        ? stems.map(
            (stem) =>
              (stem ? `${stem}${suffix}` : suffix) + (i === stringGroups.length - 1 ? '' : ':')
          )
        : [`${suffix}:`]
    );
    results.push(newResults);
  }
  return results.slice(minDepth).flat();
}

const operators = ['<', '>', '<=', '>=']; // TODO: add "none"? remove >=, <=?

/**
 * Generates all the possible suggested keywords for the given filter
 */
export function generateSuggestionsForFilter(filterDefinition: FilterDefinition) {
  const { suggestions, keywords } = filterDefinition;
  const thisFilterKeywords = Array.isArray(keywords) ? keywords : [keywords];

  const nestedSuggestions = suggestions === undefined ? [] : [suggestions];

  switch (filterDefinition.format) {
    case 'query':
      return expandStringCombinations([thisFilterKeywords, ...nestedSuggestions]);
    case 'freeform':
      return expandStringCombinations([thisFilterKeywords, []]);
    case 'range':
      return expandStringCombinations([thisFilterKeywords, ...nestedSuggestions, operators]);
    case 'rangeoverload':
      return [
        ...expandStringCombinations([thisFilterKeywords, operators]),
        ...expandStringCombinations([thisFilterKeywords, ...nestedSuggestions]),
      ];
    default:
      // Pass minDepth 1 to not generate "is:" and "not:" suggestions
      return expandStringCombinations([['is', 'not'], thisFilterKeywords], 1);
  }
}

/**
 * Return whether the query is completely valid - syntactically, and where every term matches a known filter.
 */
export function validateQuery(query: QueryAST, searchConfig: SearchConfig) {
  if (query.error) {
    return false;
  }
  switch (query.op) {
    case 'filter': {
      let filterName = query.type;
      const filterValue = query.args;

      // "is:" filters are slightly special cased
      if (filterName == 'is') {
        filterName = filterValue;
      }

      const filterDef = searchConfig.filters[filterName];
      if (filterDef) {
        // TODO: validate that filterValue is correct
        return true;
      } else {
        return false;
      }
    }
    case 'not':
      return validateQuery(query.operand, searchConfig);
    case 'and':
    case 'or': {
      return query.operands.every((q) => validateQuery(q, searchConfig));
    }
    case 'noop':
      return true;
  }
}
