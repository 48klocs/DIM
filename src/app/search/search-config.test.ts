import _ from 'lodash';
import { energyCapacityTypeNames } from './d2-known-values';
import { FilterDefinition } from './filter-types';
import { buildSearchConfig, generateSuggestionsForFilter } from './search-config';
import { allStatNames, searchableArmorStatNames } from './search-filter-values';

describe('generateSuggestionsForFilter', () => {
  const cases: [
    format: FilterDefinition['format'],
    keywords: FilterDefinition['keywords'],
    suggestions: FilterDefinition['suggestions']
  ][] = [
    [undefined, ['a', 'b', 'c'], undefined],
    ['query', 'a', ['b', 'c']],
    ['range', 'a', ['b', 'c']],
    ['range', 'a', undefined],
    ['rangeoverload', 'a', ['b', 'c']],
    ['freeform', 'a', ['b', 'c']],
    [undefined, ['a'], undefined],
    ['range', 'stat', allStatNames],
    ['query', 'maxstatvalue', searchableArmorStatNames],
    ['query', 'maxstatvalue', searchableArmorStatNames],
    ['rangeoverload', 'energycapacity', energyCapacityTypeNames],
  ];

  test.each(cases)(
    "full suggestions for filter format '%s', keyword '%s' with suggestions %s",
    (format: FilterDefinition['format'], keywords: string, suggestions?: string[]) => {
      const candidates = generateSuggestionsForFilter({
        format,
        keywords,
        suggestions,
        description: '',
        filter: () => _.stubTrue,
      });
      expect(candidates).toMatchSnapshot();
    }
  );
});

test('buildSearchConfig generates a reasonable filter map', () => {
  const searchConfig = buildSearchConfig(2);
  expect(Object.keys(searchConfig.filters)).toMatchSnapshot();
});
