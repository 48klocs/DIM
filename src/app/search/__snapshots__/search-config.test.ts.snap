// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildSearchConfig generates a reasonable filter map 1`] = `
Array [
  "dupe",
  "seasonaldupe",
  "dupelower",
  "count",
  "notes",
  "name",
  "description",
  "perk",
  "perkname",
  "keyword",
  "tagged",
  "tag",
  "hasnotes",
  "common",
  "uncommon",
  "rare",
  "legendary",
  "exotic",
  "white",
  "green",
  "blue",
  "purple",
  "yellow",
  "special",
  "primary",
  "heavy",
  "titan",
  "hunter",
  "warlock",
  "cosmetic",
  "light",
  "haslight",
  "haspower",
  "breaker",
  "kinetic",
  "arc",
  "solar",
  "void",
  "engrams",
  "lostitems",
  "messages",
  "specialorders",
  "class",
  "energy",
  "power",
  "seasonalartifacts",
  "helmet",
  "gauntlets",
  "chest",
  "leg",
  "classitem",
  "ghost",
  "clanbanners",
  "vehicle",
  "ships",
  "emblems",
  "finishers",
  "consumables",
  "modifications",
  "shaders",
  "autorifle",
  "handcannon",
  "pulserifle",
  "scoutrifle",
  "fusionrifle",
  "sniperrifle",
  "shotgun",
  "machinegun",
  "rocketlauncher",
  "sidearm",
  "sword",
  "grenadelauncher",
  "tracerifle",
  "linearfusionrifle",
  "submachine",
  "smg",
  "bow",
  "transmat",
  "weaponmod",
  "armormod",
  "reptoken",
  "powerfulreward",
  "source",
  "inloadout",
  "stack",
  "year",
  "level",
  "powerlimit",
  "masterwork",
  "energycapacity",
  "season",
  "sunsetsafter",
  "hascapacity",
  "armor2.0",
  "weapon",
  "armor",
  "equipment",
  "equippable",
  "postmaster",
  "inpostmaster",
  "equipped",
  "transferable",
  "movable",
  "stackable",
  "infusable",
  "infuse",
  "locked",
  "unlocked",
  "new",
  "randomroll",
  "curated",
  "shaded",
  "hasshader",
  "ornamented",
  "hasornament",
  "hasmod",
  "modded",
  "modslot",
  "holdsmod",
  "stat",
  "basestat",
  "maxstatloadout",
  "maxstatvalue",
  "maxbasestatvalue",
  "maxpower",
  "inleftchar",
  "inmiddlechar",
  "inrightchar",
  "onwrongclass",
  "invault",
  "incurrentchar",
  "id",
  "hash",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'freeform', keyword 'a' with suggestions [ 'b', 'c' ] 1`] = `
Array [
  "a:",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'query', keyword 'a' with suggestions [ 'b', 'c' ] 1`] = `
Array [
  "a:",
  "a:b",
  "a:c",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'query', keyword 'maxstatvalue' with suggestions [
  'mobility',  'resilience',
  'recovery',  'discipline',
  'intellect', 'strength',
  'total',     'custom',
  'any'
] 1`] = `
Array [
  "maxstatvalue:",
  "maxstatvalue:mobility",
  "maxstatvalue:resilience",
  "maxstatvalue:recovery",
  "maxstatvalue:discipline",
  "maxstatvalue:intellect",
  "maxstatvalue:strength",
  "maxstatvalue:total",
  "maxstatvalue:custom",
  "maxstatvalue:any",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'query', keyword 'maxstatvalue' with suggestions [
  'mobility',  'resilience',
  'recovery',  'discipline',
  'intellect', 'strength',
  'total',     'custom',
  'any'
] 2`] = `
Array [
  "maxstatvalue:",
  "maxstatvalue:mobility",
  "maxstatvalue:resilience",
  "maxstatvalue:recovery",
  "maxstatvalue:discipline",
  "maxstatvalue:intellect",
  "maxstatvalue:strength",
  "maxstatvalue:total",
  "maxstatvalue:custom",
  "maxstatvalue:any",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'range', keyword 'a' with suggestions [ 'b', 'c' ] 1`] = `
Array [
  "a:",
  "a:b:",
  "a:c:",
  "a:b:<",
  "a:c:<",
  "a:b:>",
  "a:c:>",
  "a:b:<=",
  "a:c:<=",
  "a:b:>=",
  "a:c:>=",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'range', keyword 'a' with suggestions undefined 1`] = `
Array [
  "a:",
  "a:<",
  "a:>",
  "a:<=",
  "a:>=",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'range', keyword 'stat' with suggestions [
  'rpm',             'rof',
  'charge',          'impact',
  'handling',        'range',
  'stability',       'reload',
  'magazine',        'aimassist',
  'equipspeed',      'velocity',
  'blastradius',     'recoildirection',
  'drawtime',        'zoom',
  'inventorysize',   'swingspeed',
  'guardefficiency', 'guardresistance',
  'chargerate',      'guardendurance',
  'ammocapacity',    'mobility',
  'resilience',      'recovery',
  'discipline',      'intellect',
  'strength',        'total',
  'custom',          'any'
] 1`] = `
Array [
  "stat:",
  "stat:rpm:",
  "stat:rof:",
  "stat:charge:",
  "stat:impact:",
  "stat:handling:",
  "stat:range:",
  "stat:stability:",
  "stat:reload:",
  "stat:magazine:",
  "stat:aimassist:",
  "stat:equipspeed:",
  "stat:velocity:",
  "stat:blastradius:",
  "stat:recoildirection:",
  "stat:drawtime:",
  "stat:zoom:",
  "stat:inventorysize:",
  "stat:swingspeed:",
  "stat:guardefficiency:",
  "stat:guardresistance:",
  "stat:chargerate:",
  "stat:guardendurance:",
  "stat:ammocapacity:",
  "stat:mobility:",
  "stat:resilience:",
  "stat:recovery:",
  "stat:discipline:",
  "stat:intellect:",
  "stat:strength:",
  "stat:total:",
  "stat:custom:",
  "stat:any:",
  "stat:rpm:<",
  "stat:rof:<",
  "stat:charge:<",
  "stat:impact:<",
  "stat:handling:<",
  "stat:range:<",
  "stat:stability:<",
  "stat:reload:<",
  "stat:magazine:<",
  "stat:aimassist:<",
  "stat:equipspeed:<",
  "stat:velocity:<",
  "stat:blastradius:<",
  "stat:recoildirection:<",
  "stat:drawtime:<",
  "stat:zoom:<",
  "stat:inventorysize:<",
  "stat:swingspeed:<",
  "stat:guardefficiency:<",
  "stat:guardresistance:<",
  "stat:chargerate:<",
  "stat:guardendurance:<",
  "stat:ammocapacity:<",
  "stat:mobility:<",
  "stat:resilience:<",
  "stat:recovery:<",
  "stat:discipline:<",
  "stat:intellect:<",
  "stat:strength:<",
  "stat:total:<",
  "stat:custom:<",
  "stat:any:<",
  "stat:rpm:>",
  "stat:rof:>",
  "stat:charge:>",
  "stat:impact:>",
  "stat:handling:>",
  "stat:range:>",
  "stat:stability:>",
  "stat:reload:>",
  "stat:magazine:>",
  "stat:aimassist:>",
  "stat:equipspeed:>",
  "stat:velocity:>",
  "stat:blastradius:>",
  "stat:recoildirection:>",
  "stat:drawtime:>",
  "stat:zoom:>",
  "stat:inventorysize:>",
  "stat:swingspeed:>",
  "stat:guardefficiency:>",
  "stat:guardresistance:>",
  "stat:chargerate:>",
  "stat:guardendurance:>",
  "stat:ammocapacity:>",
  "stat:mobility:>",
  "stat:resilience:>",
  "stat:recovery:>",
  "stat:discipline:>",
  "stat:intellect:>",
  "stat:strength:>",
  "stat:total:>",
  "stat:custom:>",
  "stat:any:>",
  "stat:rpm:<=",
  "stat:rof:<=",
  "stat:charge:<=",
  "stat:impact:<=",
  "stat:handling:<=",
  "stat:range:<=",
  "stat:stability:<=",
  "stat:reload:<=",
  "stat:magazine:<=",
  "stat:aimassist:<=",
  "stat:equipspeed:<=",
  "stat:velocity:<=",
  "stat:blastradius:<=",
  "stat:recoildirection:<=",
  "stat:drawtime:<=",
  "stat:zoom:<=",
  "stat:inventorysize:<=",
  "stat:swingspeed:<=",
  "stat:guardefficiency:<=",
  "stat:guardresistance:<=",
  "stat:chargerate:<=",
  "stat:guardendurance:<=",
  "stat:ammocapacity:<=",
  "stat:mobility:<=",
  "stat:resilience:<=",
  "stat:recovery:<=",
  "stat:discipline:<=",
  "stat:intellect:<=",
  "stat:strength:<=",
  "stat:total:<=",
  "stat:custom:<=",
  "stat:any:<=",
  "stat:rpm:>=",
  "stat:rof:>=",
  "stat:charge:>=",
  "stat:impact:>=",
  "stat:handling:>=",
  "stat:range:>=",
  "stat:stability:>=",
  "stat:reload:>=",
  "stat:magazine:>=",
  "stat:aimassist:>=",
  "stat:equipspeed:>=",
  "stat:velocity:>=",
  "stat:blastradius:>=",
  "stat:recoildirection:>=",
  "stat:drawtime:>=",
  "stat:zoom:>=",
  "stat:inventorysize:>=",
  "stat:swingspeed:>=",
  "stat:guardefficiency:>=",
  "stat:guardresistance:>=",
  "stat:chargerate:>=",
  "stat:guardendurance:>=",
  "stat:ammocapacity:>=",
  "stat:mobility:>=",
  "stat:resilience:>=",
  "stat:recovery:>=",
  "stat:discipline:>=",
  "stat:intellect:>=",
  "stat:strength:>=",
  "stat:total:>=",
  "stat:custom:>=",
  "stat:any:>=",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'rangeoverload', keyword 'a' with suggestions [ 'b', 'c' ] 1`] = `
Array [
  "a:",
  "a:<",
  "a:>",
  "a:<=",
  "a:>=",
  "a:",
  "a:b",
  "a:c",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'rangeoverload', keyword 'energycapacity' with suggestions [ 'any', 'arc', 'solar', 'void' ] 1`] = `
Array [
  "energycapacity:",
  "energycapacity:<",
  "energycapacity:>",
  "energycapacity:<=",
  "energycapacity:>=",
  "energycapacity:",
  "energycapacity:any",
  "energycapacity:arc",
  "energycapacity:solar",
  "energycapacity:void",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'undefined', keyword '[ 'a' ]' with suggestions undefined 1`] = `
Array [
  "is:a",
  "not:a",
]
`;

exports[`generateSuggestionsForFilter full suggestions for filter format 'undefined', keyword '[ 'a', 'b', 'c' ]' with suggestions undefined 1`] = `
Array [
  "is:a",
  "not:a",
  "is:b",
  "not:b",
  "is:c",
  "not:c",
]
`;
